### DISCUSSION TITLE
What is the best way to run laravel unit tests before the deploy

### DISCUSSION THREAD

**Discussion by ps-20x** [serversideup/docker-php#285] [Q&A] [answered]
I am currently not sure on what the best way is to run the unit tests. On one side I would like to do it as a first step but the containers are not done at that point.

What I want
* Run unit tests in a branch as soon as a file is pushed
* Run unit tests as soon as branch is merged to `main`
* Do not run follow-up-commands if unit tests fail

I would be really thankful if you could guide me in the right direction.

**Comment by jaydrogers** [comment, marked as answer]
Here is a good example to page through. It might not answer everything you're looking for, but it will show you an example how to get started: https://github.com/serversideup/financial-freedom/tree/main/.github/workflows

See how I have "actions" and "services"? I do that so I can reuse my pipeline code.

See [action_pr-tests.yml](https://github.com/serversideup/financial-freedom/blob/main/.github/workflows/action_pr-tests.yml) and then how it calls [service_install-and-test-php.yml](https://github.com/serversideup/financial-freedom/blob/main/.github/workflows/service_install-and-test-php.yml).

My other actions reuse the code components too. Hope this helps!

**Reply by ps-20x** [reply]
Thanks for your quick reply and showing me an example. That's perfect. I just needed to create a `.env.testing` file and create an app-key. Thank you 🙇 

One thing that I noticed on your project is:
The `action_pr-tests.yml` is referencing `./.github/workflows/service_tests-phpunit.yml` which I actually can't find. Should that actually reference your `service_install-and-test-php.yml`?
